src/app/modules/spcp/spcp.util.ts

```diff
@@ -8,13 +9,66 @@ import { MissingFeatureError } from '../core/core.errors'
 import {
   CreateRedirectUrlError,
   FetchLoginPageError,
-  InvalidAuthTypeError,
   LoginPageValidationError,
   VerifyJwtError,
 } from './spcp.errors'
 import { JwtName, JwtPayload, SpcpCookies } from './spcp.types'
 
 const logger = createLoggerWithLabel(module)
+const destinationRegex = /^\/([\w]+)\/?/
+
+const isArtifactValid = function (
+  idpPartnerEntityId: string,
+  samlArt: string,
+): boolean {
+  // Artifact should be 44 bytes long, of type 0x0004 and
+  // source id should be SHA-1 hash of the issuer's entityID
+  const hexEncodedArtifact = Buffer.from(samlArt, 'base64').toString('hex')
+  const artifactHexLength = hexEncodedArtifact.length
+  const typeCode = parseInt(hexEncodedArtifact.substr(0, 4))
+  const sourceId = hexEncodedArtifact.substr(8, 40)
+  const hashedEntityId = crypto
+    .createHash('sha1')
+    .update(idpPartnerEntityId, 'utf8')
+    .digest('hex')
+
+  return (
+    artifactHexLength === 88 && typeCode === 4 && sourceId === hashedEntityId
+  )
+}
+
+export const isValidAuthenticationQuery = (
+  samlArt: string,
+  destination: string,
+  idpPartnerEntityId: string,
+): boolean => {
+  return (
+    !!destination &&
+    isArtifactValid(idpPartnerEntityId, samlArt) &&
+    destinationRegex.test(destination)
+  )
+}
+
+export const extractFormId = (destination: string): string => {
+  // Assume that the destination has already been validated
+  const regexSplit = destinationRegex.exec(destination)!
```

Comment:
even with assumption probably good to guard against it, (and to avoid non-null assertion)